<h1 id="introduction">Introduction</h1>
<p><em>MeDeCom</em> is an R-package for reference-free decomposition of heterogeneous DNA methylation profiles. It uses matrix factorization enhanced by constraints and a specially tailored regularization. <em>MeDeCom</em> represents an input <span class="math"><em>m</em> × <em>n</em></span> data matrix (<span class="math"><em>m</em></span> CpGs measured in <span class="math"><em>n</em></span> samples) as a product of two other matrices. The first matrix has <span class="math"><em>m</em></span> rows, just as the input data, but the number of columns is equal to <span class="math"><em>k</em></span>. The columns of this matrix can be interpreted as methylomes of the <span class="math"><em>k</em></span> unknown cell populations underlying the samples and will be referred to as <strong>latent methylation components</strong> or <strong>LMCs</strong>. The second matrix has <span class="math"><em>k</em></span> rows and <span class="math"><em>n</em></span> columns, and can be interpreted as a matrix of relative contributions (mixing proportions) of each LMC to each sample.</p>
<p><em>MeDeCom</em> starts with a set of related DNA methylation profiles, e.g. a series of Infinium microarray measurements from a population cohort, or several bisulfite sequencing-based methylomes. The key requirement is that the input data represents absolute DNA methylation measurements in a population of cells and contains values between 0 and 1. <em>MeDeCom</em> implements an alternating scheme which iteratively updates randomly initialized factor matrices until convergence or until the maximum number of iterations has been reached. This is repeated for multiple random initializations and the best solution is returned.</p>
<p>MeDeCom features two tunable parameters. The first one is the number of LMCs <span class="math"><em>k</em></span>, an approximate choice for which should be known from prior information. To enforce the distribution properties of a methylation profile upon LMCs <em>MeDeCom</em> uses a special for of regularization controlled by the parameter <span class="math"><em>λ</em></span>. A typical <em>MeDeCom</em> experiment includes testing a grid of values for <span class="math"><em>k</em></span> and <span class="math"><em>λ</em></span>. For each combination of parameter values <em>MeDeCom</em> estimates a cross-validation error. The latter helps select the optimal number of LMCs and the strength of regularization.</p>
<h1 id="installation">Installation</h1>
<p><em>MeDeCom</em> can be installed directly from github using the package <code>devtools</code>:</p>
<pre class="sourceCode r"><code class="sourceCode r">devtools:::<span class="kw">install_github</span>(<span class="st">&quot;lutsik/MeDeCom&quot;</span>)</code></pre>
<p>Currently only *nix-like platforms with a C++11-compatible compiler are supported. MeDeCom uses stack model for memory to accelerate factorization for smaller ranks. This requires certain preparation during compilation, therefore, please, note the extended compilation time (15 to 20 minutes).</p>
<h1 id="data-preparation">Data preparation</h1>
<p><em>MeDeCom</em> accepts DNA methylation data in several forms. Preferably the user may load and preprocess the data using a general-purpose DNA methylation analysis package <a href="http://rnbeads.mpi-inf.mpg.de">RnBeads</a>. A resulting RnBSet object can be directly supplied to <em>MeDeCom</em>. Alternatively, <em>MeDeCom</em> runs on any matrix of type <code>numeric</code> with valid methylation values.</p>
<p><em>MeDeCom</em> comes with a small example data set obtained by mixing reference profiles of blood cell methylomes <em>in silico</em>. The example data set can be loaded in a usual way:</p>
<pre class="sourceCode r"><code class="sourceCode r">## load the package
<span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(MeDeCom))
##  load the example data sets
<span class="kw">data</span>(example.dataset, <span class="dt">package=</span><span class="st">&quot;MeDeCom&quot;</span>)
## you should get objects D, Tref and Aref
## in your global R environment
<span class="kw">ls</span>()</code></pre>
<pre><code>## [1] &quot;Aref&quot;           &quot;D&quot;              &quot;Tref&quot;           &quot;lmcs&quot;          
## [5] &quot;medecom.result&quot; &quot;perm&quot;           &quot;prop&quot;           &quot;sample.group&quot;  
## [9] &quot;sge.setup&quot;</code></pre>
<p>Loaded numeric matrix <code>D</code> contains 100 <em>in silico</em> mixtures and serves as an example input. Columns of matrix <code>Tref</code> contains the methylomes of 5 blood cell types used to generate the mixtures, while matrix <code>Aref</code> provides the mixing proportions.</p>
<pre class="sourceCode r"><code class="sourceCode r">## matrix D has dimension 10000x100
<span class="kw">str</span>(D)</code></pre>
<pre><code>##  num [1:10000, 1:100] 0.0354 0.0491 0.3411 0.873 0.1781 ...</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">## matrix Tref has dimension 10000x5
<span class="kw">str</span>(Tref)</code></pre>
<pre><code>##  num [1:10000, 1:5] 0.1 0.0847 0.3167 0.834 0.0455 ...
##  - attr(*, &quot;dimnames&quot;)=List of 2
##   ..$ : NULL
##   ..$ : chr [1:5] &quot;Neutrophils&quot; &quot;CD4+ T-cells&quot; &quot;CD14+ Monocytes&quot; &quot;CD8+ T-cells&quot; ...</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">## matrix Aref has dimension 5x100
<span class="kw">str</span>(Aref)</code></pre>
<pre><code>##  num [1:5, 1:100] 6.54e-01 3.80e-02 3.06e-01 1.56e-13 2.45e-03 ...</code></pre>
<h1 id="performing-a-methylome-decomposition-experiment">Performing a methylome decomposition experiment</h1>
<p><em>MeDeCom</em> can be run directly on matrix <code>D</code>.</p>
<p>It is crucial to select the values of parameters <span class="math"><em>k</em></span> and <span class="math"><em>λ</em></span> to test. A choice of <span class="math"><em>k</em></span> is often dictated by prior knowledge about the methylomes. Precise value of lambda has to be selected for each data set independently. A good start is a logarithmic grid of lambda values. It is important to include <span class="math"><em>λ</em> = 0</span> into the grid, as this particular case the regularization is effectively absent making <em>MeDeCom</em> similar to other NMF-based deconvolution algorithms.</p>
<pre><code>medecom.result&lt;-runMeDeCom(D, Ks=2:10, lambdas=c(0,10^(-5:-1)))</code></pre>
<p><em>MeDeCom</em> is based upon an alternating optimization heuristic and requires a lot of computation. The processing of the data matrix can take several hours. One can speed up the run by decreasing the number of cross-validation folds and random initializations, and increasing the number of computational cores.</p>
<pre class="sourceCode r"><code class="sourceCode r">medecom.result&lt;-<span class="kw">runMeDeCom</span>(D, <span class="dv">2</span>:<span class="dv">10</span>, <span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">10</span>^(-<span class="dv">5</span>:-<span class="dv">1</span>)), <span class="dt">NINIT=</span><span class="dv">10</span>, <span class="dt">NFOLDS=</span><span class="dv">10</span>, <span class="dt">ITERMAX=</span><span class="dv">300</span>, <span class="dt">NCORES=</span><span class="dv">9</span>)</code></pre>
<pre><code>## 
## [Main:] checking inputs
## [Main:] preparing data
## [Main:] preparing jobs
## [Main:] 3114 factorization runs in total
## [Main:] runs 2755 to 2788 complete
## [Main:] runs 2789 to 2822 complete
## [Main:] runs 2823 to 2856 complete
## [Main:] runs 2857 to 2890 complete
## ......
## [Main:] finished all jobs. Creating the object</code></pre>
<p>This can, however, lead to decomposition slightly different from the one presented given below.</p>
<p>The results of a decomposition experiment are saved to an object of class <code>MeDeComSet</code>. The contents of an object can be conveniently displayed using the <code>print</code> functionality.</p>
<pre class="sourceCode r"><code class="sourceCode r">medecom.result</code></pre>
<pre><code>## An object of class MeDeComSet
## Input data set:
##  10000 CpGs
##  100 methylomes
## Experimental parameters:
##  k values: 2, 3, 4, 5, 6, 7, 8, 9, 10
##  lambda values: 0, 1e-05, 1e-04, 0.001, 0.01, 0.1</code></pre>
<h1 id="exploring-the-decomposition-results">Exploring the decomposition results</h1>
<h2 id="parameter-selection">Parameter selection</h2>
<p>The first key step is parameter selection. It is important to carefully explore the obtained results and make a decision about the most feasible parameter values, or about extending the parameter value grids to be tested in refinement experiments.</p>
<p><em>MeDeCom</em> provides a <strong>cross-validation error</strong> (CVE) for each tested parameter combination.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotParameters</span>(medecom.result)</code></pre>
<div class="figure">
<img src="figure/unnamed-chunk-7-1.png" alt="plot of chunk unnamed-chunk-7" /><p class="caption">plot of chunk unnamed-chunk-7</p>
</div>
<p>A lineplot helping to select parameter <span class="math"><em>λ</em></span> can be produced by specifying a fixed value for <span class="math"><em>k</em></span>:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotParameters</span>(medecom.result, <span class="dt">K=</span><span class="dv">5</span>, <span class="dt">lambdaScale=</span><span class="st">&quot;log&quot;</span>)</code></pre>
<div class="figure">
<img src="figure/unnamed-chunk-8-1.png" alt="plot of chunk unnamed-chunk-8" /><p class="caption">plot of chunk unnamed-chunk-8</p>
</div>
<p>Cross-validation error has a minimum at <span class="math"><em>λ</em> = 10<sup> − 2</sup></span> so this value is preferred.</p>
<h2 id="latent-methylation-components-lmcs">Latent methylation components (LMCs)</h2>
<p>A matrix of LMCs can be extracted using <code>getLMCs</code>:</p>
<pre class="sourceCode r"><code class="sourceCode r">lmcs&lt;-<span class="kw">getLMCs</span>(medecom.result, <span class="dt">K=</span><span class="dv">5</span>, <span class="dt">lambda=</span><span class="fl">0.01</span>)
<span class="kw">str</span>(lmcs)</code></pre>
<pre><code>##  num [1:10000, 1:5] 0 0.0184 0.2898 0.7856 0 ...</code></pre>
<p>LMCs can be seen as measured methylation profiles of purified cell populations. <em>MeDeCom</em> provides for several visualization methods for LMCs using the function <code>plotLMCs</code> which operates directly on <code>MeDeComSet</code> objects.</p>
<h3 id="clustering">Clustering</h3>
<p>For instance, standard hierarchical clustering can be visualized using:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotLMCs</span>(medecom.result, <span class="dt">K=</span><span class="dv">5</span>, <span class="dt">lambda=</span><span class="fl">0.01</span>, <span class="dt">type=</span><span class="st">&quot;dendrogram&quot;</span>)</code></pre>
<div class="figure">
<img src="figure/unnamed-chunk-10-1.png" alt="plot of chunk unnamed-chunk-10" /><p class="caption">plot of chunk unnamed-chunk-10</p>
</div>
<p>A two-dimensional embedding with MDS is also obtainable:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotLMCs</span>(medecom.result, <span class="dt">K=</span><span class="dv">5</span>, <span class="dt">lambda=</span><span class="fl">0.01</span>, <span class="dt">type=</span><span class="st">&quot;MDS&quot;</span>)</code></pre>
<div class="figure">
<img src="figure/unnamed-chunk-11-1.png" alt="plot of chunk unnamed-chunk-11" /><p class="caption">plot of chunk unnamed-chunk-11</p>
</div>
<p>Input data can be included into the MDS plot to enhance the interpretation.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotLMCs</span>(medecom.result, <span class="dt">K=</span><span class="dv">5</span>, <span class="dt">lambda=</span><span class="fl">0.01</span>, <span class="dt">type=</span><span class="st">&quot;MDS&quot;</span>, <span class="dt">D=</span>D)</code></pre>
<div class="figure">
<img src="figure/unnamed-chunk-12-1.png" alt="plot of chunk unnamed-chunk-12" /><p class="caption">plot of chunk unnamed-chunk-12</p>
</div>
<h3 id="matching-lmcs-to-reference-profiles">Matching LMCs to reference profiles</h3>
<p>In many cases reference methylomes exists, which are relevant for the data set in question. For our example analysis matrix <code>Tref</code> contains the reference type profiles which were <em>in silico</em> mixed. <em>MeDeCom</em> offers several ways to visualize the resulting LMCs together with the reference methylation profiles. The reference methylomes can be included into a joint clustering analysis:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotLMCs</span>(medecom.result, <span class="dt">K=</span><span class="dv">5</span>, <span class="dt">lambda=</span><span class="fl">0.01</span>, <span class="dt">type=</span><span class="st">&quot;dendrogram&quot;</span>, <span class="dt">Tref=</span>Tref, <span class="dt">center=</span><span class="ot">TRUE</span>)</code></pre>
<div class="figure">
<img src="figure/unnamed-chunk-13-1.png" alt="plot of chunk unnamed-chunk-13" /><p class="caption">plot of chunk unnamed-chunk-13</p>
</div>
<p>Furthermore, a similarity matrix of LMCs vs reference profiles can be visualized as a heatmap.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotLMCs</span>(medecom.result, <span class="dt">K=</span><span class="dv">5</span>, <span class="dt">lambda=</span><span class="fl">0.01</span>, <span class="dt">type=</span><span class="st">&quot;heatmap&quot;</span>, <span class="dt">Tref=</span>Tref)</code></pre>
<div class="figure">
<img src="figure/unnamed-chunk-14-1.png" alt="plot of chunk unnamed-chunk-14" /><p class="caption">plot of chunk unnamed-chunk-14</p>
</div>
<p>Correlation coefficient values and asterisks aid the interpretation. The values are displayed in the cells which contain maximal values column-wise. The asterisks mark cells which have the highest correlation value in the respective rows. Thus, a value with asterisk corresponds to a mutual match, i.e. LMC unambiguously matching a reference profile.</p>
<p>In this example analysis each LMC uniquely matches one of the reference profiles. The matching of</p>
<p>Function <code>matchLMCs</code> offers several methods for matching LMCs to reference profiles.</p>
<pre class="sourceCode r"><code class="sourceCode r">perm&lt;-<span class="kw">matchLMCs</span>(lmcs, Tref)</code></pre>
<h3 id="lmc-enrichment-analysis">LMC enrichment analysis</h3>
<p>MeDeCom provides functions to perform enrichment analysis on the sites that are particularly hypo-/hypermethylated in an LMC. These sites can then be used for GO and LOLA enrichment analysis. Importantly, genomic annotations of the LMC sites is required to be specified. We thus recommend to use the DecompPipelie <a href="https://github.com/lutsik/DecompPipeline">https://github.com/lutsik/DecompPipeline</a> for processing, but the annotation can also be specified manually using a <code>data.frame</code> that looks as follows:</p>
<pre class="sourceCode r"><code class="sourceCode r">      Chromosome   Start     End Strand CpG GC CGI Relation SNPs
<span class="dv">30365</span>       chr1 <span class="dv">1036375</span> <span class="dv">1036376</span>      +<span class="st">   </span><span class="dv">2</span> <span class="dv">59</span>        Shelf &lt;<span class="ot">NA</span>&gt;
<span class="dv">42681</span>       chr1 <span class="dv">1184537</span> <span class="dv">1184538</span>      +<span class="st">   </span><span class="dv">2</span> <span class="dv">58</span>     Open Sea &lt;<span class="ot">NA</span>&gt;
<span class="dv">45091</span>       chr1 <span class="dv">1218625</span> <span class="dv">1218626</span>      +<span class="st">   </span><span class="dv">5</span> <span class="dv">66</span>       Island &lt;<span class="ot">NA</span>&gt;
<span class="dv">51615</span>       chr1 <span class="dv">1292773</span> <span class="dv">1292774</span>      +<span class="st">   </span><span class="dv">3</span> <span class="dv">64</span>       Island &lt;<span class="ot">NA</span>&gt;
<span class="dv">52001</span>       chr1 <span class="dv">1295504</span> <span class="dv">1295505</span>      +<span class="st">   </span><span class="dv">9</span> <span class="dv">65</span>        Shore &lt;<span class="ot">NA</span>&gt;
<span class="dv">52003</span>       chr1 <span class="dv">1295507</span> <span class="dv">1295508</span>      +<span class="st">   </span><span class="dv">9</span> <span class="dv">65</span>        Shore &lt;<span class="ot">NA</span>&gt;</code></pre>
<p>The required columns are <code>Chromosome</code>, <code>Start</code>, <code>End</code>, and <code>Strand</code>. Using this <code>data.frame</code> (called <code>df</code> in the following), enrichment analysis can be performed using:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">lmc.lola.enrichment</span>(medecom.result,<span class="dt">anno.data=</span>df,<span class="dt">K=</span><span class="dv">5</span>,<span class="dt">lambda=</span><span class="fl">0.001</span>,<span class="dt">diff.threshold =</span> <span class="fl">0.5</span>, <span class="dt">region.type =</span> <span class="st">&quot;tiling&quot;</span>)</code></pre>
<p>Please note that <code>df</code> needs to have the same number of rows than the methylation matrix used as input to MeDeCom. CpGs are first aggregated over the <code>region.type</code> specified, then the regions are selected that have a difference larger than <code>diff.threshold</code>. The list of available region types is published here <a href="https://rnbeads.org/regions.htm">https://rnbeads.org/regions.html</a>.</p>
<h2 id="mixing-proportions">Mixing proportions</h2>
<p>A matrix of mixing proportions is obtained using <code>getProportions</code>:</p>
<pre class="sourceCode r"><code class="sourceCode r">prop&lt;-<span class="kw">getProportions</span>(medecom.result, <span class="dt">K=</span><span class="dv">5</span>, <span class="dt">lambda=</span><span class="fl">0.001</span>)
<span class="kw">str</span>(prop)</code></pre>
<pre><code>##  num [1:5, 1:100] 0.3411 0 0.065 0.0213 0.5726 ...
##  - attr(*, &quot;dimnames&quot;)=List of 2
##   ..$ : chr [1:5] &quot;LMC1&quot; &quot;LMC2&quot; &quot;LMC3&quot; &quot;LMC4&quot; ...
##   ..$ : NULL</code></pre>
<h3 id="visualization-of-the-complete-proportion-matrix">Visualization of the complete proportion matrix</h3>
<p>A complete matrix of propotions can be visualized as a stacked barplot:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotProportions</span>(medecom.result, <span class="dt">K=</span><span class="dv">5</span>, <span class="dt">lambda=</span><span class="fl">0.01</span>, <span class="dt">type=</span><span class="st">&quot;barplot&quot;</span>)</code></pre>
<div class="figure">
<img src="figure/unnamed-chunk-17-1.png" alt="plot of chunk unnamed-chunk-17" /><p class="caption">plot of chunk unnamed-chunk-17</p>
</div>
<p>or a heatmap:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotProportions</span>(medecom.result, <span class="dt">K=</span><span class="dv">5</span>, <span class="dt">lambda=</span><span class="fl">0.01</span>, <span class="dt">type=</span><span class="st">&quot;heatmap&quot;</span>)</code></pre>
<div class="figure">
<img src="figure/unnamed-chunk-18-1.png" alt="plot of chunk unnamed-chunk-18" /><p class="caption">plot of chunk unnamed-chunk-18</p>
</div>
<p>The heatmap can be enhanced by clustering the columns:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotProportions</span>(medecom.result, <span class="dt">K=</span><span class="dv">5</span>, <span class="dt">lambda=</span><span class="fl">0.01</span>, <span class="dt">type=</span><span class="st">&quot;heatmap&quot;</span>, <span class="dt">heatmap.clusterCols=</span><span class="ot">TRUE</span>)</code></pre>
<div class="figure">
<img src="figure/unnamed-chunk-19-1.png" alt="plot of chunk unnamed-chunk-19" /><p class="caption">plot of chunk unnamed-chunk-19</p>
</div>
<p>or adding color code for the samples:</p>
<pre class="sourceCode r"><code class="sourceCode r">sample.group&lt;-<span class="kw">c</span>(<span class="st">&quot;Case&quot;</span>, <span class="st">&quot;Control&quot;</span>)[<span class="dv">1</span>+<span class="kw">sample.int</span>(<span class="kw">ncol</span>(D))%%<span class="dv">2</span>]
<span class="kw">plotProportions</span>(medecom.result, <span class="dt">K=</span><span class="dv">5</span>, <span class="dt">lambda=</span><span class="fl">0.01</span>, <span class="dt">type=</span><span class="st">&quot;heatmap&quot;</span>, <span class="dt">sample.characteristic=</span>sample.group)</code></pre>
<div class="figure">
<img src="figure/unnamed-chunk-20-1.png" alt="plot of chunk unnamed-chunk-20" /><p class="caption">plot of chunk unnamed-chunk-20</p>
</div>
<h3 id="visualization-of-selected-lmc-proportions">Visualization of selected LMC proportions</h3>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotProportions</span>(medecom.result,  <span class="dt">K=</span><span class="dv">5</span>, <span class="dt">lambda=</span><span class="fl">0.01</span>, <span class="dt">type=</span><span class="st">&quot;lineplot&quot;</span>, <span class="dt">lmc=</span><span class="dv">2</span>, <span class="dt">Aref=</span>Aref, <span class="dt">ref.profile=</span><span class="dv">2</span>)</code></pre>
<div class="figure">
<img src="figure/unnamed-chunk-22-1.png" alt="plot of chunk unnamed-chunk-22" /><p class="caption">plot of chunk unnamed-chunk-22</p>
</div>
<h1 id="advanced-usage">Advanced usage</h1>
<h2 id="running-medecom-on-a-compute-cluster">Running <em>MeDeCom</em> on a compute cluster</h2>
<p><em>MeDeCom</em> experiments require a lot of computational time. On the other hand most of the factorization runs are independent and, therefore, can be run in parallel. Thus, a significant speedup can be achieved when running <em>MeDeCom</em> in an HPC environment. <em>MeDeCom</em> can be easily adapted to most of the popular schedulers. There are, however, several prerequisites:</p>
<ul>
<li>the scheduler provides the standard utilities <code>qsub</code> for the submission of the cluster jobs and <code>qstat</code> for obtaining the job statistics;</li>
<li>the cluster does not have a low limit on the number of submitted jobs;</li>
<li>the R installation (location of the R binary and the package library) is consistent across the execution nodes.</li>
</ul>
<p>The example below is for the cluster operated by <em>Son of Grid Engine</em> (SoGE). To be able to run on a SoGE cluster <em>MeDeCom</em> needs to know:</p>
<ul>
<li>location of the R executable (directory);</li>
<li>an operating memory limit per each factorization job;</li>
<li>a pattern for the names of cluster nodes to run the jobs on.</li>
</ul>
<p>These settings should be stored in a <code>list</code> object:</p>
<pre class="sourceCode r"><code class="sourceCode r">sge.setup&lt;-<span class="kw">list</span>(
<span class="dt">R_bin_dir=</span><span class="st">&quot;/usr/bin&quot;</span>,
<span class="dt">host_pattern=</span><span class="st">&quot;*&quot;</span>,
<span class="dt">mem_limit=</span><span class="st">&quot;5G&quot;</span>
)</code></pre>
<p>This object should be supplied to <em>MeDeCom</em> as the argument <code>cluster.settings</code>. It is also important to specify a valid temporary directory, which is available to all execution nodes.</p>
<pre class="sourceCode r"><code class="sourceCode r">medecom.result&lt;-<span class="kw">runMeDeCom</span>(D, <span class="dt">Ks=</span><span class="dv">2</span>:<span class="dv">10</span>, <span class="dt">lambdas=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">10</span>^(-<span class="dv">5</span>:-<span class="dv">1</span>)), <span class="dt">N_COMP_LAMBDA=</span><span class="dv">1</span>, <span class="dt">NFOLDS=</span><span class="dv">5</span>, <span class="dt">NINIT=</span><span class="dv">10</span>, 
<span class="dt">temp.dir=</span><span class="st">&quot;/cluster_fs/medecom_temp&quot;</span>,
<span class="dt">cluster.settings=</span>sge.setup)</code></pre>
<p><em>MeDeCom</em> will start the jobs and will periodically monitor the number of remaining ones.</p>
<pre><code>## 
## [Main:] checking inputs
## [Main:] preparing data
## [Main:] preparing jobs
## [Main:] 3114 factorization runs in total
## [Main:] 3114 jobs remaining
## ....
## [Main:] finished all jobs. Creating the object</code></pre>
<h1 id="r-session">R session</h1>
<p>Here is the output of <code>sessionInfo()</code> on the system on which this document was compiled:</p>
<pre><code>## R version 3.6.0 (2019-04-26)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Debian GNU/Linux 7 (wheezy)
## 
## Matrix products: default
## BLAS:   /TL/deep-share/archive00/software/packages/R/R-3.6.0/lib/libRblas.so
## LAPACK: /TL/deep-share/archive00/software/packages/R/R-3.6.0/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_US       LC_NUMERIC=C         LC_TIME=en_US       
##  [4] LC_COLLATE=C         LC_MONETARY=en_US    LC_MESSAGES=en_US   
##  [7] LC_PAPER=en_US       LC_NAME=C            LC_ADDRESS=C        
## [10] LC_TELEPHONE=C       LC_MEASUREMENT=en_US LC_IDENTIFICATION=C 
## 
## attached base packages:
##  [1] grid      stats4    parallel  stats     graphics  grDevices utils    
##  [8] datasets  methods   base     
## 
## other attached packages:
##  [1] MeDeCom_0.3.0                          
##  [2] RnBeads_2.3.2                          
##  [3] plyr_1.8.4                             
##  [4] methylumi_2.30.0                       
##  [5] minfi_1.30.0                           
##  [6] bumphunter_1.26.0                      
##  [7] locfit_1.5-9.1                         
##  [8] iterators_1.0.12                       
##  [9] foreach_1.4.7                          
## [10] Biostrings_2.52.0                      
## [11] XVector_0.24.0                         
## [12] SummarizedExperiment_1.14.0            
## [13] DelayedArray_0.10.0                    
## [14] BiocParallel_1.18.0                    
## [15] FDb.InfiniumMethylation.hg19_2.2.0     
## [16] org.Hs.eg.db_3.8.2                     
## [17] TxDb.Hsapiens.UCSC.hg19.knownGene_3.2.2
## [18] GenomicFeatures_1.36.0                 
## [19] AnnotationDbi_1.46.0                   
## [20] reshape2_1.4.3                         
## [21] scales_1.0.0                           
## [22] Biobase_2.44.0                         
## [23] illuminaio_0.26.0                      
## [24] matrixStats_0.54.0                     
## [25] limma_3.40.0                           
## [26] gridExtra_2.3                          
## [27] ggplot2_3.1.1                          
## [28] fields_9.7                             
## [29] maps_3.3.0                             
## [30] spam_2.2-2                             
## [31] dotCall64_1.0-0                        
## [32] ff_2.2-14                              
## [33] bit_1.1-14                             
## [34] cluster_2.0.8                          
## [35] MASS_7.3-51.4                          
## [36] GenomicRanges_1.36.0                   
## [37] GenomeInfoDb_1.20.0                    
## [38] IRanges_2.18.0                         
## [39] S4Vectors_0.22.0                       
## [40] BiocGenerics_0.30.0                    
## [41] RUnit_0.4.32                           
## [42] gplots_3.0.1.1                         
## [43] gtools_3.8.1                           
## [44] pracma_2.2.5                           
## [45] Rcpp_1.0.3                             
## [46] knitr_1.22                             
## 
## loaded via a namespace (and not attached):
##  [1] colorspace_1.4-1         siggenes_1.58.0         
##  [3] mclust_5.4.3             base64_2.0              
##  [5] bit64_0.9-7              xml2_1.2.0              
##  [7] splines_3.6.0            codetools_0.2-16        
##  [9] scrime_1.3.5             zeallot_0.1.0           
## [11] Rsamtools_2.0.0          annotate_1.62.0         
## [13] HDF5Array_1.12.1         readr_1.3.1             
## [15] compiler_3.6.0           httr_1.4.0              
## [17] backports_1.1.5          assertthat_0.2.1        
## [19] Matrix_1.2-17            lazyeval_0.2.2          
## [21] prettyunits_1.0.2        tools_3.6.0             
## [23] gtable_0.3.0             glue_1.3.1              
## [25] GenomeInfoDbData_1.2.1   dplyr_0.8.0.1           
## [27] doRNG_1.7.1              vctrs_0.2.0             
## [29] multtest_2.40.0          nlme_3.1-139            
## [31] preprocessCore_1.46.0    gdata_2.18.0            
## [33] rtracklayer_1.44.0       DelayedMatrixStats_1.6.0
## [35] xfun_0.11                stringr_1.4.0           
## [37] rngtools_1.3.1.1         XML_3.98-1.19           
## [39] beanplot_1.2             zlibbioc_1.30.0         
## [41] hms_0.4.2                GEOquery_2.52.0         
## [43] rhdf5_2.28.0             RColorBrewer_1.1-2      
## [45] memoise_1.1.0            pkgmaker_0.27           
## [47] biomaRt_2.40.0           reshape_0.8.8           
## [49] stringi_1.4.3            RSQLite_2.1.2           
## [51] highr_0.8                genefilter_1.66.0       
## [53] caTools_1.17.1.2         bibtex_0.4.2            
## [55] rlang_0.4.1              pkgconfig_2.0.3         
## [57] bitops_1.0-6             nor1mix_1.2-3           
## [59] evaluate_0.14            lattice_0.20-38         
## [61] purrr_0.3.2              Rhdf5lib_1.6.0          
## [63] GenomicAlignments_1.20.0 tidyselect_0.2.5        
## [65] magrittr_1.5             R6_2.4.0                
## [67] DBI_1.0.0                pillar_1.3.1            
## [69] withr_2.1.2              survival_2.44-1.1       
## [71] RCurl_1.95-4.12          tibble_2.1.1            
## [73] crayon_1.3.4             KernSmooth_2.23-15      
## [75] progress_1.2.0           data.table_1.12.6       
## [77] blob_1.2.0               digest_0.6.22           
## [79] xtable_1.8-4             tidyr_0.8.3             
## [81] openssl_1.3              munsell_0.5.0           
## [83] registry_0.5-1           quadprog_1.5-7          
## [85] askpass_1.1</code></pre>
